src/Backups/BackupImpl.h

```diff
@@ -1,92 +1,107 @@
+    SizeAndChecksum getFileSizeAndChecksum(const String & file_name) const override;
     BackupEntryPtr readFile(const String & file_name) const override;
+    BackupEntryPtr readFile(const SizeAndChecksum & size_and_checksum) const override;
     void writeFile(const String & file_name, BackupEntryPtr entry) override;
     void finalizeWriting() override;
-
-protected:
-    /// Checks if this backup exists.
-    virtual bool backupExists() const = 0;
-
-    virtual void openImpl(OpenMode open_mode_) = 0;
-    OpenMode getOpenModeNoLock() const { return open_mode; }
-
-    virtual void closeImpl(const Strings & written_files_, bool writing_finalized_) = 0;
-
-    /// Read a file from the backup.
-    /// Low level: the function doesn't check base backup or checksums.
-    virtual std::unique_ptr<ReadBuffer> readFileImpl(const String & file_name) const = 0;
-
-    /// Add a file to the backup.
-    /// Low level: the function doesn't check base backup or checksums.
-    virtual std::unique_ptr<WriteBuffer> writeFileImpl(const String & file_name) = 0;
-
-    mutable std::mutex mutex;
+    bool supportsWritingInMultipleThreads() const override { return !use_archives; }
 
 private:
+    using FileInfo = IBackupCoordination::FileInfo;
+    class BackupEntryFromBackupImpl;
+
+    void open();
+    void close();
     void writeBackupMetadata();
     void readBackupMetadata();
-
-    struct FileInfo
-    {
-        UInt64 size = 0;
-        UInt128 checksum{0, 0};
-
-        /// for incremental backups
-        UInt64 base_size = 0;
-        UInt128 base_checksum{0, 0};
-    };
-
-    class BackupEntryFromBackupImpl;
+    String getArchiveNameWithSuffix(const String & suffix) const;
+    std::shared_ptr<IArchiveReader> getArchiveReader(const String & suffix) const;
+    std::shared_ptr<IArchiveWriter> getArchiveWriter(const String & suffix);
+    void removeAllFilesAfterFailure();
 
     const String backup_name;
+    const ArchiveParams archive_params;
+    const bool use_archives;
+    const OpenMode open_mode;
+    std::shared_ptr<IBackupWriter> writer;
+    std::shared_ptr<IBackupReader> reader;
+    const bool is_internal_backup;
+    std::shared_ptr<IBackupCoordination> coordination;
     ContextPtr context;
-    const std::optional<BackupInfo> base_backup_info_param;
-    OpenMode open_mode = OpenMode::NONE;
-    UUID uuid = {};
+
+    mutable std::mutex mutex;
+    std::optional<UUID> uuid;
     time_t timestamp = 0;
+    UInt64 version;
     std::optional<BackupInfo> base_backup_info;
     std::shared_ptr<const IBackup> base_backup;
     std::optional<UUID> base_backup_uuid;
-    std::map<String, FileInfo> file_infos; /// Should be ordered alphabetically, see listFiles().
-    std::unordered_map<UInt128, String> file_checksums;
-    Strings written_files;
+    mutable std::unordered_map<String /* archive_suffix */, std::shared_ptr<IArchiveReader>> archive_readers;
+    std::pair<String, std::shared_ptr<IArchiveWriter>> archive_writers[2];
```

Comment:
This is because we need to keep no more than two archive writers at any moment: we add data files to the current archive, and we need to keep a writer for the first archive in order to be able at the end of the backuping to write the backup's metadata to the first archive.

Reply:
Anyway I've disabled my implementation of multiple archives for a single backup, I think I'll rework it later